### ISSUE TITLE
Run envoy pod as non-root

### ISSUE THREAD

**Issue by mkilchhofer** [projectcontour/contour#2907]
**Please describe the problem you have**
Referring to the blog post at [kubernetes.io > 11 Ways (Not) to Get Hacked](https://kubernetes.io/blog/2018/07/18/11-ways-not-to-get-hacked/#8-run-containers-as-a-non-root-user) we should try running envoy as a non root user.

**Comment by youngnick** [comment]
Yes, you can definitely do that - it's a little confusing though because you have to tell *Contour* that Envoy should be configured on 8080 and 8443 as well as set up the Daemonset and Service wiring.

Contour will actually default to using those ports for Envoy's HTTP and HTTPS listener respectively, in our example YAMLs, we override that using `--envoy-service-http-port` and --envoy-service-https-port`.

That will finish up the network parts, and allow the requests to flow correctly. I'll be curious to know if everything works after that.

Once you're all done, please consider writing this up as a Guide in `site/_guides`, it would be great to have more documentation of this.

**Comment by mkilchhofer** [comment]
Hi @youngnick,
Thanks for explaining it. I knew I already saw these params while I was working on the bitnami helm chart improvements for contour.
I already have a working non-root setup with the bitnami chart: https://github.com/bitnami/charts/pull/3674.

Regarding writing this down in a guide: I will do it as far as I find some time to do this :+1:

**Comment by binnythomas-1989** [comment]
Guys I tried the same using @mkilchhofer  helm chart changes that he made. Had to do some more stuff mentioned here https://github.com/bitnami/charts/pull/3674 to get it working.  But The envoy container on 8443 says connection refused.

```
I have no name!@contour-ingress-envoy-csbdw:/$ curl localhost:8080
I have no name!@contour-ingress-envoy-csbdw:/$ curl localhost:8443
curl: (7) Failed to connect to localhost port 8443: Connection refused
```

**Comment by youngnick** [comment]
Contour won't open either the secure or insecure ports unless you have something that needs it. Do you have a Ingress or HTTPPRoxy object that requests TLS configured @binnythomas-1989?

**Comment by binnythomas-1989** [comment]
Yes this  is sorted @youngnick  It only opens one when we created an ingress resource. Can you check this https://github.com/projectcontour/contour/issues/2922

**Comment by jpeach** [comment]
xref https://github.com/kubernetes/kubernetes/issues/56374

**Comment by embano1** [comment]
Had a similar use case and created this patch:

```
diff --git a/examples/contour/03-envoy.yaml b/examples/contour/03-envoy.yaml
index 9265b74a..2e0283d9 100644
--- a/examples/contour/03-envoy.yaml
+++ b/examples/contour/03-envoy.yaml
@@ -23,6 +23,8 @@ spec:
       labels:
         app: envoy
     spec:
+      hostNetwork: true
+      dnsPolicy: ClusterFirstWithHostNet
       containers:
       - command:
         - /bin/contour
@@ -51,6 +53,9 @@ spec:
         - --service-cluster $(CONTOUR_NAMESPACE)
         - --service-node $(ENVOY_POD_NAME)
         - --log-level info
+        securityContext:
+          runAsUser: 101
+          allowPrivilegeEscalation: false
         command:
         - envoy
         image: docker.io/envoyproxy/envoy:v1.16.1
@@ -104,6 +109,9 @@ spec:
         - --envoy-cafile=/certs/ca.crt
         - --envoy-cert-file=/certs/tls.crt
         - --envoy-key-file=/certs/tls.key
+        securityContext:
+          runAsUser: 101
+          allowPrivilegeEscalation: false
         command:
         - contour
         image: docker.io/projectcontour/contour:main
```

Ignore the `hostNetwork` stuff as it's just an example for deploying in non-LB environments but still running as non-root as one workaround to avoid [CVE-2020–15257](https://medium.com/nttlabs/dont-use-host-network-namespace-f548aeeef575).

I can file a PR for these changes if it makes sense.

**Comment by youngnick** [comment]
I'd love for Contour to be confirmed able to run as non-root, but I'm concerned about interactions with the standard installs. Does this interact with the currently-default hostPort setup? Or if we swap to NodePort?

Aside from that small concern, I'd love to see this in - running as root is not great, and remapping the userns by default seems much better. This would certainly need a changelog note though.
